File: android/src/main/kotlin/com/neelakandan/flutter_neo_shield/location/LocationIntegrityChecker.kt
package com.neelakandan.flutter_neo_shield.location

/**
 * Layer 7: Environment Integrity Check.
 *
 * Aggregates scores from all detection layers using weighted scoring.
 * Cross-references results for consistency and applies amplification.
 */
class LocationIntegrityChecker {

    private val weights = mapOf(
        "mockProvider" to 1.0,
        "spoofingApp" to 0.9,
        "locationHook" to 0.95,
        "gpsSignal" to 0.7,
        "sensorFusion" to 0.8,
        "temporalAnomaly" to 0.85
    )

    /**
     * Layers whose positive result is conclusive on its own. A single such
     * signal (e.g. an active mock-location provider) must flag as spoofed even
     * though the weighted average of all six layers would fall below the 0.5
     * threshold — otherwise a real spoof using only mock location goes
     * undetected while the "advanced" layers report 0.
     */
    private val vetoThresholds = mapOf(
        "mockProvider" to 1.0,
        "locationHook" to 0.95,
        "spoofingApp" to 1.0
    )

    /** Confidence assigned when a high-confidence veto layer fires. */
    private val vetoConfidence = 0.9

    /** Compute weighted confidence from all layer scores. */
    fun computeConfidence(scores: Map<String, Double>): Double {
        var totalScore = 0.0
        var totalWeight = 0.0

        for ((key, weight) in weights) {
            val score = scores[key] ?: 0.0
            totalScore += score * weight
            totalWeight += weight
        }

        if (totalWeight == 0.0) return 0.0

        val normalized = totalScore / totalWeight

        // Cross-validation amplification:
        // If multiple layers agree, increase confidence
        val triggeredLayers = scores.count { it.value > 0.3 }
        val amplifier = when {
            triggeredLayers >= 4 -> 1.5
            triggeredLayers >= 3 -> 1.3
            triggeredLayers >= 2 -> 1.1
            else -> 1.0
        }

        val weighted = (normalized * amplifier).coerceIn(0.0, 1.0)

        // High-confidence veto: any definitive single signal is conclusive.
        val vetoed = vetoThresholds.any { (key, threshold) ->
            (scores[key] ?: 0.0) >= threshold
        }

        return if (vetoed) maxOf(weighted, vetoConfidence) else weighted
    }
}

File: ios/Classes/location/LocationIntegrityChecker.swift
import Foundation

/// Layer 7: Environment Integrity Check for iOS.
class LocationIntegrityChecker {

    private let weights: [String: Double] = [
        "mockProvider": 1.0,
        "spoofingApp": 0.9,
        "locationHook": 0.95,
        "gpsSignal": 0.7,
        "sensorFusion": 0.8,
        "temporalAnomaly": 0.85,
    ]

    /// Layers whose positive result is conclusive on its own. A single such
    /// signal (e.g. a simulated location) must flag as spoofed even though the
    /// weighted average of all six layers would fall below the 0.5 threshold.
    private let vetoThresholds: [String: Double] = [
        "mockProvider": 1.0,
        "locationHook": 0.95,
        "spoofingApp": 1.0,
    ]

    /// Confidence assigned when a high-confidence veto layer fires.
    private let vetoConfidence: Double = 0.9

    func computeConfidence(scores: [String: Double]) -> Double {
        var totalScore: Double = 0.0
        var totalWeight: Double = 0.0

        for (key, weight) in weights {
            let score = scores[key] ?? 0.0
            totalScore += score * weight
            totalWeight += weight
        }

        guard totalWeight > 0 else { return 0.0 }

        let normalized = totalScore / totalWeight

        let triggeredLayers = scores.filter { $0.value > 0.3 }.count
        let amplifier: Double
        switch triggeredLayers {
        case 4...: amplifier = 1.5
        case 3: amplifier = 1.3
        case 2: amplifier = 1.1
        default: amplifier = 1.0
        }

        let weighted = min(normalized * amplifier, 1.0)

        // High-confidence veto: any definitive single signal is conclusive.
        let vetoed = vetoThresholds.contains { key, threshold in
            (scores[key] ?? 0.0) >= threshold
        }

        return vetoed ? max(weighted, vetoConfidence) : weighted
    }
}
